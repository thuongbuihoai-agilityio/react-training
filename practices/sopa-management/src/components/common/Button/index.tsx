import {
  ButtonHTMLAttributes,
  memo
} from 'react';

// Styles
import './button.css';

export enum ButtonType {
  default = '',
  primary = 'primary',
  secondary = 'secondary',
  tertiary = 'tertiary',
  btnIconPrimary = 'icon-primary',
  btnIconSecondary = 'icon-secondary',
  btnOutlinePrimary = 'out-line-primary',
  btnOutlineSecondary = 'out-line-secondary',
}

interface ButtonProps {
  ariaLabel?: string;
  children?: string | JSX.Element;
  type?: ButtonType;
  disable?: boolean;
  submit?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
  className?: string;
  onClick?: () => void;
  onSubmit?: () => void;
}

const Button = ({
  ariaLabel = '',
  children = '',
  type = ButtonType.default,
  disable = false,
  submit,
  className = '',
  onClick,
  onSubmit
}: ButtonProps) => (
  <button
    data-testid='button'
    aria-label={ariaLabel}
    className={`${className} btn btn-${type}`}
    disabled={disable}
    type={submit}
    onClick={onClick}
    onSubmit={onSubmit}
  >
    {children}
  </button>
);

export default memo(Button);
