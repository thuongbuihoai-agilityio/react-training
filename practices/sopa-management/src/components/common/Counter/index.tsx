import { memo } from 'react';

// Components
import {
  Button,
  ButtonType,
  Input
} from '@components/common';

// Styles
import './counter.css';

interface CounterProps {
  value: number;
  onIncrement: () => void;
  onDecrement: () => void;
};

const Counter = ({
  value,
  onIncrement,
  onDecrement
}: CounterProps) => {
  const isDisable = value === 1;

  return (
    <div data-testid='counter' className='counter'>
      <Button
        ariaLabel='Decrement'
        type={ButtonType.btnIconPrimary}
        className={isDisable ? 'counter-minus-disable' : 'counter-minus'}
        disable={isDisable}
        onClick={onDecrement}
      />
      <Input type='number' classNameInput='counter-input' value={value} />
      <Button
        ariaLabel='Increment'
        type={ButtonType.btnIconPrimary}
        className='counter-plus'
        onClick={onIncrement}
      />
    </div>
  );
}

export default memo(Counter);
