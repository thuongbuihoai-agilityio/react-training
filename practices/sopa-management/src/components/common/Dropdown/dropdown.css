.custom-dropdown {
  position: relative;
  display: inline-block;
  width: var(--w-medium);
  border-radius: var(--rounded-tiny);
  border: 1px solid var(--border-primary);
}

.dropdown-label,
.custom-dropdown > a[role="button"],
.custom-dropdown > .dropdown-link {
  display: inline-block;
  padding: 15px;
  text-decoration: none;
  cursor: pointer;
  border-radius: var(--rounded-tiny);
}

.dropdown-label:after,
.custom-dropdown > a[role="button"]:after {
  content: '';
  position: absolute;
  left: 335px;
  top: 40%;
  display: inline-block;
  width: 12px;
  height: 8px;
  background-image: url(../../../assets/icons/arrow.svg);
}

.dropdown-list {
  position: absolute;
  list-style: none;
  border-radius: var(--rounded-tiny);
  margin-top: var(--m-tiny);
  z-index: 1;
  width: 100%;
  background: var(--light);
  border: 1px solid var(--border-primary);
  padding: 3px 0;
}

.dropdown-option {
  display: block;
  padding: var(--p-tiny) var(--py-extra-base);
  text-decoration: none;
  color: var(--dark);
}

.dropdown-option:hover,
.dropdown-option:focus {
  background: var(--border-primary);
  cursor: pointer;
}

.dropdown-link {
  color: var(--dark);
}
