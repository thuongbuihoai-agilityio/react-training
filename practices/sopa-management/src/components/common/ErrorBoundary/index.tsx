import { Component, ReactNode } from 'react';

interface Props {
  children: ReactNode;
  fallback: ReactNode;
}

interface State {
  hasError: boolean;
}

class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(): State {
    return { hasError: true };
  }

  render() {
    const { hasError } = this.state;
    const { children, fallback } = this.props;

    if (hasError) {
      return fallback;
    }

    return children;
  }
}

export default ErrorBoundary;
