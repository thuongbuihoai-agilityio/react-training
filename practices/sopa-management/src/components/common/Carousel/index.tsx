// Images
import { IMAGE } from '@constants/image';

// Components
import {
  Button,
  ButtonType,
  Image,
  Input,
  SizeType,
  Text,
} from '@components/common';

// Constants
import { CAROUSEL_BULLET } from '@constants/common';

// Styles
import './carousel.css';

const Carousel = () => (
  <div data-testid='carousel' className='carousel'>
    <div className='carousel-inner'>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-1'
        name='carousel'
        hidden
        checked
      />
      <figure className='carousel-item'>
        <Image
          url={IMAGE.banner}
          alt={IMAGE.altBanner}
          width={1170}
          height={657}
        />
      </figure>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-2'
        name='carousel'
        hidden
      />
      <figure className='carousel-item'>
        <Image
          url={IMAGE.banner2}
          alt={IMAGE.altBanner}
          width={1170}
          height={657}
          size='50vw'
        />
      </figure>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-3'
        name='carousel'
        hidden
      />
      <figure className='carousel-item'>
        <Image
          url={IMAGE.banner}
          alt={IMAGE.altBanner}
          width={1170}
          height={657}
        />
      </figure>
      <ul className='carousel-indicators'>
        {CAROUSEL_BULLET.map((item) => (
          <li key={item.key}>
            <label htmlFor={item.htmlFor} className={item.className} />
          </li>
        ))}
      </ul>
    </div>
    <div className='carousel-info'>
      <Text
        text='Step inside, for comfort and magic await you.'
        type={SizeType.large}
        className='carousel-heading'
      />
      <Button
        ariaLabel='Show now'
        children='Show now'
        type={ButtonType.primary}
      />
    </div>
  </div>
);

export default Carousel;
