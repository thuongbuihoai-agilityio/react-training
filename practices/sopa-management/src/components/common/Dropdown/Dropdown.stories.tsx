// Libs
import type { Meta, StoryObj } from '@storybook/react';

// Components
import Dropdown from '@common/Dropdown';

// Constants
import { SIZE } from '@constants/common';

const meta: Meta<typeof Dropdown> = {
  component: Dropdown,
};

export default meta;
type Story = StoryObj<typeof Dropdown>;

export const DropdownSize: Story = {
  args: {
    data: SIZE,
  }
};
