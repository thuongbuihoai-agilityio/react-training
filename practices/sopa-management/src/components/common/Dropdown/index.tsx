import {
  memo,
  useCallback,
  useState
} from 'react';
import { Link } from 'react-router-dom';

// Components
import { Text } from '@components/common';

// Styles
import './dropdown.css';

export type SizeType = {
  key?: string;
  label?: string;
};

export type UserType = {
  key?: string;
  label?: string;
  href: string;
};

interface DropdownProps {
  data?: SizeType[];
  value?: string;
  isHref?: boolean;
  className?: string;
  onClick?: () => void;
  onSetValue?: (value?: string) => void;
}

const Dropdown = ({
  data = [],
  isHref,
  value,
  className = '',
  onClick,
  onSetValue = () => {}
}: DropdownProps) => {
  const [openDropdown, setOpenDropdown] = useState<boolean>(false);

  const handleToggle = useCallback(() => {
    setOpenDropdown(!openDropdown);
  }, [openDropdown]);

  return (
    <div
      data-testid='dropdown'
      className={`${className} custom-dropdown`}
      onClick={handleToggle}
    >
      {isHref ? (
        <Link
          to='/login'
          className='dropdown-link'
          aria-label='link to login page'
          onClick={onClick}
        >
          Logout
        </Link>
      ) : (
        <>
          <Text text={value} className='dropdown-label' />
          {openDropdown && (
            <ul className='dropdown-list'>
              {data.map((item: SizeType) => (
                <li
                  id='my-dropdown'
                  key={item.key}
                  value={item.label}
                  className='dropdown-option'
                  onClick={() => onSetValue(item.label)}
                >
                  {item.label}
                </li>
              ))}
            </ul>
          )}
        </>
      )}
    </div>
  );
};
export default memo(Dropdown);
