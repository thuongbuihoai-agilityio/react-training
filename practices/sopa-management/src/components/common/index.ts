import Button, { ButtonType } from './Button';
import Carousel from './Carousel';
import Counter from './Counter';
import Dropdown from './Dropdown';
import ErrorBoundary from './ErrorBoundary';
import Icon, { IconType } from './Icon';
import Image from './Image';
import Input, { InputTheme, InputType } from './Input';
import Loading from './Loading';
import Logout from './Logout';
import Menu, { MenuTheme } from './Menu';
import Partners from './Partners';
import PopupDelete from './PopupDelete';
import Price, { PriceType } from './Price';
import Rating from './Rating';
import RatingStar from './Rating/RatingStar';
import Social from './Social';
import Text, { ThemeType, SizeType } from './Text';

export {
  Button,
  ButtonType,
  Carousel,
  Counter,
  Dropdown,
  ErrorBoundary,
  Icon,
  IconType,
  Image,
  Input,
  InputTheme,
  InputType,
  Loading,
  Logout,
  Menu,
  MenuTheme,
  Partners,
  PopupDelete,
  Price,
  PriceType,
  Rating,
  RatingStar,
  Social,
  Text,
  ThemeType,
  SizeType
};
