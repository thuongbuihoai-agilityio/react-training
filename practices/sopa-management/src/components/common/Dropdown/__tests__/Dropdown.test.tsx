// Libs
import '@testing-library/jest-dom';
import {
  render,
  screen
} from '@testing-library/react';

// Components
import Dropdown from '@common/Dropdown';

// Constants
import { SIZE } from '@constants/common';
import { renderRouterTest } from '@helpers/testUtils';
import userEvent from '@testing-library/user-event';

describe('Dropdown component', () => {
  test('should render Dropdown component', () => {
    const { getByTestId } = render(<Dropdown data={SIZE} />);
    expect(getByTestId('dropdown')).toBeInTheDocument();
  });

  test('should render Dropdown component with default props', () => {
    const { getByTestId } = render(<Dropdown />);
    expect(getByTestId('dropdown')).toBeInTheDocument();
  });

  test('should render correctly when the dropdown is open', async () => {
    const { getByTestId } = render(<Dropdown data={SIZE} />);

    const dropdown = getByTestId('dropdown');
    await userEvent.click(dropdown);

    expect(screen.getByText('Small')).toBeInTheDocument();
    expect(screen.getByText('Medium')).toBeInTheDocument();
    expect(screen.getByText('Large')).toBeInTheDocument();
  });

  test('should render correctly when the dropdown is open with logout', async () => {
    const { getByTestId } = renderRouterTest(<Dropdown isHref data={SIZE} />);

    const dropdown = getByTestId('dropdown');
    await userEvent.click(dropdown);

    expect(screen.getByText('Logout')).toBeInTheDocument();
  });

  test('should call onSetValue with the correct value when an item is clicked', async () => {
    const mockOnSetValue = jest.fn();
    const { getByTestId } = render(<Dropdown data={SIZE} onSetValue={mockOnSetValue} />);

    const dropdown = getByTestId('dropdown');
    await userEvent.click(dropdown);

    const option1 = screen.getByText('Small');
    await userEvent.click(option1);

    expect(mockOnSetValue).toHaveBeenCalledWith('Small');
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<Dropdown data={SIZE} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
