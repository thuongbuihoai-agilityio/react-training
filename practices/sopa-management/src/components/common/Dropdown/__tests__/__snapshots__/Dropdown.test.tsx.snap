// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class=" custom-dropdown"
    data-testid="dropdown"
  >
    <p
      class="dropdown-label text-"
      data-testid="text"
    >
      Sopa
    </p>
  </div>
</DocumentFragment>
`;
